python/ray/data/dataset.py

```diff
@@ -1390,8 +1390,13 @@ def duplicate_row(row: Dict[str, Any]) -> List[Dict[str, Any]]:
     @PublicAPI(api_group=BT_API_GROUP)
     def filter(
         self,
-        fn: Optional[UserDefinedFunction[Dict[str, Any], bool]] = None,
-        expr: Optional[str] = None,
+        fn: Optional[
+            UserDefinedFunction[Dict[str, Any], bool]
+        ] = None,  # TODO: Deprecate this parameter in favor of predicate
+        expr: Optional[
+            str
+        ] = None,  # TODO: Deprecate this parameter in favor of predicate
+        predicate: Optional[Expr] = None,
```

Comment:
Not a fan of polymorphism, but much less of a fan of messing up our public APIs.

 - Let's instead pass it as `expr` param
 - Update docs that the strings are deprecated and will be removed in the future
 - Print warning if expr is str